libs/pika/execution/include/pika/execution/algorithms/when_all_vector.hpp

```diff
@@ -0,0 +1,323 @@
+//  Copyright (c) 2020 ETH Zurich
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <pika/config.hpp>
+#include <pika/assert.hpp>
+#include <pika/concepts/concepts.hpp>
+#include <pika/datastructures/optional.hpp>
+#include <pika/datastructures/variant.hpp>
+#include <pika/execution/algorithms/detail/single_result.hpp>
+#include <pika/execution_base/operation_state.hpp>
+#include <pika/execution_base/receiver.hpp>
+#include <pika/execution_base/sender.hpp>
+#include <pika/functional/detail/tag_fallback_invoke.hpp>
+#include <pika/type_support/detail/with_result_of.hpp>
+#include <pika/type_support/pack.hpp>
+
+#include <atomic>
+#include <cstddef>
+#include <exception>
+#include <functional>
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace pika::execution::experimental {
+    namespace when_all_vector_detail {
+        template <typename Sender>
+        struct when_all_vector_sender_impl
+        {
+            struct when_all_vector_sender_type;
+        };
+
+        template <typename Sender>
+        using when_all_vector_sender = typename when_all_vector_sender_impl<
+            Sender>::when_all_vector_sender_type;
+
+        template <typename Sender>
+        struct when_all_vector_sender_impl<Sender>::when_all_vector_sender_type
+        {
+            using senders_type = std::vector<Sender>;
+            senders_type senders;
+
+            explicit constexpr when_all_vector_sender_type(
+                senders_type&& senders)
+              : senders(PIKA_MOVE(senders))
+            {
+                PIKA_ASSERT_MSG(this->senders.size() >= 1,
+                    "when_all_vector expects at least one element in the "
+                    "vector of senders");
```

Comment:
To do: this leaves it up to the caller to check that the vector isn't empty. Is that necessary?

Reply:
I'm fine with leaving it out